net/ping/ping.go

```diff
@@ -0,0 +1,233 @@
+	// applies to all future I/O, so we only need to do it once.
+	deadline, ok := ctx.Deadline()
+	if ok {
+		if err := p.c.SetReadDeadline(deadline); err != nil {
+			return nil, err
+		}
+	}
+
+	p.wg.Add(1)
+	go p.run(ctx)
+	return p, nil
+}
+
+func newUnstarted(ctx context.Context, logf logger.Logf) (*Pinger, error) {
+	var id [2]byte
+	_, err := rand.Read(id[:])
+	if err != nil {
+		return nil, err
+	}
+
+	conn, err := netns.Listener(logf).ListenPacket(ctx, "ip4:icmp", "0.0.0.0")
+	if err != nil {
+		return nil, err
+	}
+
+	return &Pinger{
+		c:       conn,
+		Logf:    logf,
+		timeNow: time.Now,
+		id:      uint16(id[0]) | uint16(id[1])<<8,
+		pings:   make(map[uint16]chan response),
+	}, nil
+}
+
+func (p *Pinger) logf(format string, a ...any) {
+	if p.Logf != nil {
+		p.Logf(format, a...)
+	} else {
+		log.Printf(format, a...)
+	}
+}
+
+func (p *Pinger) vlogf(format string, a ...any) {
+	if p.Verbose {
+		p.logf(format, a...)
+	}
+}
+
+func (p *Pinger) Close() error {
+	err := p.c.Close()
+	p.wg.Wait()
+	return err
+}
+
+func (p *Pinger) run(ctx context.Context) {
+	defer p.wg.Done()
+	buf := make([]byte, 1500)
+
+loop:
+	for {
+		select {
+		case <-ctx.Done():
+			break loop
+		default:
+		}
+
+		n, addr, err := p.c.ReadFrom(buf)
+		if err != nil {
+			// Ignore temporary errors; everything else is fatal
+			if netErr, ok := err.(net.Error); !ok || !netErr.Temporary() {
+				break
+			}
+			continue
+		}
+
+		p.handleResponse(buf[:n], addr, p.timeNow())
+	}
+
+	p.cleanupOutstanding()
+}
+
+func (p *Pinger) cleanupOutstanding() {
+	// Complete outstanding requests
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	for _, ch := range p.pings {
+		ch <- response{err: net.ErrClosed}
+	}
+}
+
+func (p *Pinger) handleResponse(buf []byte, addr net.Addr, now time.Time) {
+	const ProtocolICMP = 1
+	m, err := icmp.ParseMessage(ProtocolICMP, buf)
+	if err != nil {
+		p.vlogf("handleResponse: invalid packet: %v", err)
+		return
+	}
+
+	if m.Type != ipv4.ICMPTypeEchoReply {
+		p.vlogf("handleResponse: wanted m.Type=%d; got %d", ipv4.ICMPTypeEchoReply, m.Type)
+		return
+	}
+
+	resp, ok := m.Body.(*icmp.Echo)
+	if !ok || resp == nil {
+		p.vlogf("handleResponse: wanted body=*icmp.Echo; got %v", m.Body)
+		return
+	}
+
+	// We assume we sent this if the ID in the response is ours.
+	if uint16(resp.ID) != p.id {
+		p.vlogf("handleResponse: wanted ID=%d; got %d", p.id, resp.ID)
+		return
+	}
+
+	// Search for existing running echo request
+	var ch chan response
+	p.mu.Lock()
+	if ch, ok = p.pings[uint16(resp.Seq)]; ok {
+		delete(p.pings, uint16(resp.Seq))
+	} else {
+		p.vlogf("handleResponse: got response for unknown Seq %d", resp.Seq)
+	}
+	p.mu.Unlock()
+
+	if ch != nil {
+		ch <- response{t: now}
+	}
+}
+
+func (p *Pinger) Send(ctx context.Context, dest net.Addr) (time.Duration, error) {
+	// Use sequential sequence numbers on the assumption that we will not
+	// wrap around when using a single Pinger instance
+	p.mu.Lock()
+	p.seq++
```

Comment:
optional: a random sequence might help with the ID space issue

Reply:
I'm thinking of colliding with unrelated pings originating from the same host. They tend to start at 0 so if we don't, we're more likely to miss those.

Actually, we'd greatly reduce the risk of collisions with pings from other processes by checking that the reply's source IP matches where we sent the ping to.

Reply:
... the hostname of the derp server in the payload probably suffices to distinguish the responses anyway